### DISCUSSION TITLE
counter and rps

### DISCUSSION THREAD

**Discussion by bellad17** [micropython/micropython#15454] [RP2040 / Pico]
hello, i want with rp2040 do a counter and tachometer ,
 i have a slide gate , i put on gear pinion shaft a magnets ,
 and for control my slide gate with hall effect and rp2040
control 1 : count when hall effect go in front magnet ( open or close gate slide ) if overcounting there is an action
control 2 : revolutions per second ( open or close ) for if slowing down there is an action
but it turns slowly
i try with  , but no good , my hall effect it's omnipolaire
can you help , thank you
```py
# import the require libraries
from machine import Pin
import utime

# global variable
counter = 0

# constants
SAMPLING_TIME = 1 # ( in seconds )
                  # You may also ask the user to input sample time

# pin declaration
tachometerPin = Pin(13, Pin.IN, Pin.PULL_UP)


# interrupt handler function
def tachometer(pin):# pin is default positional argument
    global counter
    counter += 1
    
# attach the interrupt to the tachometer Pin
tachometerPin.irq(trigger = Pin.IRQ_FALLING,
                  handler = tachometer)


# main logic/ function of the program
while True:
    utime.sleep(SAMPLING_TIME)
    
    revolutions_per_sampling_time = counter / 2 # two white strips on the rotor
    revolutions_per_sec = revolutions_per_sampling_time / SAMPLING_TIME
    revolutions_per_minute = revolutions_per_sec * 60
    
    print("RPM : ", revolutions_per_minute )
    # reset the counter to zero 
    counter = 0
```

**Comment by bixb922** [comment]
What problem are you having?
How many pulses per second are you expecting?
What hall sensor are you using?

**Comment by bellad17** [comment]
my problem is print 1 or 0 no more
i have try print("RPM : ", revolutions_per_sec ) same
my sensor TLE4913 
it turns really slowly

**Reply by scruss** [reply]
> ... my sensor TLE4913

As @bixb922 already noted, this isn't the right sensor for your application. The [datasheet](https://www.infineon.com/dgdl/Infineon-TLE4913-DS-v02_03-EN.pdf?fileId=5546d46269e1c019016ac0287fa13265) shows it only samples for 50 µs about 8 times per second.

The TLE4913 works well for its original application — powering up a flip-phone when the user opens it — but isn't what you want here.

**Reply by bellad17** [reply]
which sensorr should I use ?

**Reply by bellad17** [reply]
the SS461A it's ok ??? 
[datasheet](https://www.mouser.fr/datasheet/2/187/HWSC_S_A0012720386_1-3073244.pdf)

**Comment by bixb922** [comment]
I don't understand well what the problem is, sorry. These are possible problems:

AFAIK the TLE4913 samples the input only once every 100 milliseconds. It may miss pulses if they are shorter than this time. 

In the time between reading the `counter` variable and setting it to zero, there may be an interrupt and it will be missed. This should not occur often, since that time window is very short. You could compute the difference with the last value of `counter` instead.

**Comment by bellad17** [comment]
can be used with timer hardware , but I don't know how
i'hv exemple but in basic
``'bas
Dim Flag220 as Bit  'flag indicating less that 220 rpm
Dim Flag6000 as Bit  'flag indicationg less than 6000 rpm
Dim Gate_in_action as bit  'must be set to 1 when gate is moving
Config timer0 = timer, prescaler = 1024
Stop timer0
On timer0 ISR_220rpm  'timer-overflow at speeds below 220 rpm
enable timer0
Config int0 = rising
On int0 ISR_Hall
enable int0
enable interrupts
Do
If Gate_in_action = 1 start timer0 else stop timer0
If flag6000 = 1 then
 flag6000 = 0
 "Action A"  'whatever it is
End if
If flag220 = 1 then
 flag220 = 0
 gate_in_action = 0
 "Action B"  'whatever it is
End if
Loop
ISR_Hall:
 If timer0 > 10 then flag6000 = 1  'more than 10 ms between hall signals, i.e. less than 6000 rpm
 timer0 = 0
return
IS_220rpm
 flag220 = 1
 stop timer0  'to prevent that he ISR is executed every 260 milliseconds
Return
'''
